### ISSUE TITLE
Handling of javascript: navigations is not interoperable, spec doesn't match most implementations

### ISSUE THREAD

**Issue by bzbarsky** [w3c/webappsec-csp#322]
Consider this testcase.  Parent page:

    <iframe src="subframe.html"></iframe>
    <div onclick="frames[0].location.href = 'javascript:alert(5)">
      Click me
    </div>

subframe.html:

    <meta http-equiv="content-security-policy" content="script-src 'self'">

When the user clicks the link, what should happen?

In Firefox there is an alert.  In Chrome, Safari, and Edge there is no alert.

Consider also this testcase.   Parent page:

    <meta http-equiv="content-security-policy" content="script-src 'self'">
    <iframe src="subframe.html"></iframe>
    <div>
      Click me
    </div>
    <script src="test.js"></script>

test.js:

    document.querySelector('div').onclick = function() { frames[0].location.href = 'javascript:alert(5)' }

and subframe.html has no CSP.  Now when the user clicks the link, what should happen?  Now Firefox does not have an alert, while Chrome, Safari, and Edge do.

Going to the spec, we are landing in https://w3c.github.io/webappsec-csp/#should-block-navigation-request with _source_ the parent page and _target_ the subframe.  The check only considers _source_'s CSP list, which is consistent with the behavior of Firefox but not any of the other browsers here.

@mikewest why is Chrome implementing something totally different from the spec in this case?  Did the spec just get mis-written or something and should be using _target_'s CSP for the unsafe-inline check?

**Comment by bzbarsky** [comment]
Oh, and the _source_ is the parent because https://html.spec.whatwg.org/multipage/history.html#location-object-navigate says that comes from the incumbent settings on location sets.  The other obvious way to have _source_ != _target_ here would be to use a targeted `<a>` or `<form>`, but Chrome doesn't support that at all, so I can't test that behavior.

**Comment by bzbarsky** [comment]
And there are no web platform tests at all for this behavior, as far as I can tell.  WPT's content-security-policy/navigation/to-javascript-url-script-src.html only tests .src on iframes, as  does content-security-policy/navigation/to-javascript-url-frame-src.html

**Comment by andypaicu** [comment]
It seems that chrome has fallen in line since this issue was raised, the spec is definitely correct here.

I'm adding 2 tests here: https://chromium-review.googlesource.com/c/chromium/src/+/1348054

Once it lands they will be under
content-security-policy/navigation/to-javascript-parent-initiated-child-csp.html
content-security-policy/navigation/to-javascript-parent-initiated-parent-csp.html

They are pretty much the tests in the original post.

Will close when it lands.

**Comment by dinofx** [comment]
Shouldn't one of these examples fail due to unsafe evaluation of dynamic code?  Consider just the page:

```html
<meta http-equiv="content-security-policy" content="script-src 'unsafe-inline'">
<script>
hash=document.location.hash.substring(1);
malicious='javascript:' + hash;
eval(hash);
</script>
<div onclick="window.location.href = malicious" >
  Click me
</div>
```

Open the page with the URL fragment `#alert(5)`.  Note that `eval(hash)` is prevented (using Chrome), but the script can still be evaluated by prefixing it with `javascript:` and assigning it to `window.location.href`.

**Comment by mbrodesser-Igalia** [comment]
@dinofx: `unsafe-inline` allows executing inline event handlers and `javascript:urls`, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources.

**Comment by dinofx** [comment]
@mbrodesser-Igalia the different sources are pretty clear to me.  In my example above, the JavaScript:url is not inline, it’s dynamic.

**Comment by mbrodesser-Igalia** [comment]
> @mbrodesser-Igalia the different sources are pretty clear to me. In my example above, the JavaScript:url is not inline, it’s dynamic.

AFAIU, allowing inline scripts via `script-src 'unsafe-inline'` implies those scripts may depend on user-provided data / dynamic code. That understanding is based on MDN's documentation; the spec itself might state something different which would need to be pointed out here.

**Comment by dinofx** [comment]
Inline means the script that will be evaluated was embedded in the original HTML, in its entirety.  Otherwise it would be the same as unsafe eval

**Comment by mbrodesser-Igalia** [comment]
> Inline means the script that will be evaluated was embedded in the original HTML, in its entirety.

I couldn't find a definition in the spec for "inline script"; "inline", in the spec, is used without further definition. It would add clarification to define it. CC @annevk 

> Otherwise it would be the same as unsafe eval

It'd allow executing user-defined strings. An extension of https://github.com/w3c/webappsec-csp/issues/322#issuecomment-811135901 is:

```
<meta http-equiv="content-security-policy" content="script-src 'unsafe-inline'">
<script>
hash=location.hash.substring(1);
malicious='javascript:' + hash;
eval(hash);
</script>
<div onclick="location.href = malicious" >
  Click me
</div>
<script>
document.querySelector("div").click();
</script>
```

**Comment by dinofx** [comment]
> An extension of

I'm not sure how that modified example is different.  Assigning `location.href` to a string that starts with `javascript:` is a form of `eval(...)`.  It should only execute when `unsafe-eval` is allowed.

These are both inline and equivalent (neither allows arbitrary code to execute):
```
<a href="javascript:onClick()" >click me</a>
<div onclick="onClick()"> click me</div>
```

And these two are equivalent (both are a form of eval):
```
<div onclick="location.href = 'javascript:' + malicious" >
<div onclick="eval(malicious)" >
```

The latter two should be blocked without `unsafe-eval`.

**Comment by mbrodesser-Igalia** [comment]
> > An extension of
> 
> I'm not sure how that modified example is different. Assigning `location.href` to a string that starts with `javascript:` is a form of `eval(...)`. It should only execute when `unsafe-eval` is allowed.
> 
> These are both inline and equivalent (neither allows arbitrary code to execute):
> 
> ```
> <a href="javascript:onClick()" >click me</a>
> <div onclick="onClick()"> click me</div>
> ```
> 
> And these two are equivalent (both are a form of eval):
> 
> ```
> <div onclick="location.href = 'javascript:' + malicious" >
> <div onclick="eval(malicious)" >
> ```
> 
> The latter two should be blocked without `unsafe-eval`.

@dinofx: consider filing a separate ticket for this to raise attention to it.
